crates/c-api/include/wasmtime/async.h

```diff
@@ -0,0 +1,265 @@
+ * this method will configure what happens when fuel runs out. Specifically executing
+ * WebAssembly will be suspended and control will be yielded back to the caller.
+ *
+ * This is only suitable with use of a store associated with an async config because
+ * only then are futures used and yields are possible.
+ */
+WASM_API_EXTERN void wasmtime_context_out_of_fuel_async_yield(
+    wasmtime_context_t *context,
+    uint64_t injection_count,
+    uint64_t fuel_to_inject);
+
+/**
+ * \brief Configures epoch-deadline expiration to yield to the async caller and the update the deadline.
+ *
+ * This is only suitable with use of a store associated with an async config because
+ * only then are futures used and yields are possible.
+ *
+ * See the Rust documentation for more:
+ * https://docs.wasmtime.dev/api/wasmtime/struct.Store.html#method.epoch_deadline_async_yield_and_update
+ */
+WASM_API_EXTERN void wasmtime_context_epoch_deadline_async_yield_and_update(
+    wasmtime_context_t *context,
+    uint64_t delta);
+
+/**
+ * The callback to determine a continuation's current state.
+ *
+ * Return true if the host call has completed, otherwise false will 
+ * continue to yield WebAssembly execution.
+ */
+typedef bool (*wasmtime_func_async_continuation_callback_t)(void *env);
+
+/**
+ * A continuation for the current state of the host function's execution.
+ */
+typedef struct wasmtime_async_continuation_t {
+  /// Callback for if the async function has completed.
+  wasmtime_func_async_continuation_callback_t callback;
+  /// User-provided argument to pass to the callback.
+  void *env;
+  /// A finalizer for the user-provided *env
+  void (*finalizer)(void *);
+} wasmtime_async_continuation_t;
+
+/**
+ * \brief Callback signature for #wasmtime_linker_define_async_func.
+ *
+ * This is a host function that returns a continuation to be called later. 
+ * The continuation returned is owned by wasmtime and will be deleted when it completes.
+ *
+ * All the arguments to this function will be kept alive until the continuation
+ * returns that it has errored or has completed.
+ *
+ * \param env user-provided argument passed to #wasmtime_linker_define_async_func
+ * \param caller a temporary object that can only be used during this function
+ * call. Used to acquire #wasmtime_context_t or caller's state
+ * \param args the arguments provided to this function invocation
+ * \param nargs how many arguments are provided
+ * \param results where to write the results of this function
+ * \param nresults how many results must be produced
+ * \param trap_ret if assigned a not `NULL` value then the called function will
+ *        trap with the returned error. Note that ownership of trap is transferred
+ *        to wasmtime.
+ *
+ * Only supported for async stores.
+ *
+ * See #wasmtime_func_callback_t for more information.
+ */
+typedef wasmtime_async_continuation_t *(*wasmtime_func_async_callback_t)(
```

Comment:
Instead of returning `wasmtime_async_continuation_t` could that perhaps be an out-ptr? That way the contract of this function is that it always fills in `wasmtime_async_continuation_t` and then must fill in `trap_ret` and `results` by the time the continuation yields "I'm done"?

Reply:
In lieu of this though there are other issues - for example in Rust this is represented in the return value as a `Box<T>` but that's not guaranteed to use the same allocator as `new` uses in C++. By using a retptr it would help sidestep such issues by taking allocators out of the picture.

Reply:
Ah this is why I originally had the `async_continuation_new` function... I can make that an out ptr instead, seems better and allows for removing a heap allocation.